Disable pagechange event if page is not changing.

yurydelendik opened the pull request:
Fixes #7288

yurydelendik commented:
Let's see if it breaks our viewer

/botio-windows preview

yurydelendik commented:
- [ ] check with mozcentral tests

Snuffleupagus commented:
This seems to cause some glitches related to Presentation Mode, e.g
1. Open the preview.
2. Go to the first page.
3. Enter Presentation Mode.
4. Exit Presentation Mode.

Note how the *top* of the first page is now scrolled down a bit, when compared with the same steps in http://mozilla.github.io/pdf.js/web/viewer.html.

Snuffleupagus commented:
The title of the PR is "Disable pagechange event if page is not changing.", but the code actually goes a bit further than that. If the intent of the patch is simply to not fire the event unnecessarily, then we probably still need https://github.com/mozilla/pdf.js/blob/master/web/pdf_viewer.js#L192-L196 even in the `this._currentPageNumber === val` case.
Please note that I've not tested it, but I'm guessing that that code would fix the the Presentation Mode glitch that I described above, and might also prevent other issues where existing code relies on the current behaviour.

yurydelendik commented:
> then we probably still need https://github.com/mozilla/pdf.js/blob/master/web/pdf_viewer.js#L192-L196 even in the this._currentPageNumber === val case.

That what I was afraid of. We need to remove this logic from currentPageNumber somehow.

yurydelendik commented:
I was not able to reproduce to STR above. However I agree that not calling this.scrollPageIntoView might break things. I refactored it further to remove updateInProgress and extracted _resetCurrentPageView -- now the logic must be preserved.

/botio-windows preview

yurydelendik commented on web/pdf_viewer.js:
```diff
@@ -673,13 +676,11 @@ var PDFViewer = (function pdfViewer() {
 
       if (!this.isInPresentationMode) {
-        this.currentPageNumber = currentId;
+        this._setCurrentPageNumber(currentId, true);
```
this might cause the confusion -- we need to rethink this logic in the future

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -165,12 +164,21 @@ var PDFViewer = (function pdfViewer() {
+    },
 
+    _setCurrentPageNumber:
+        function pdfViewer_setCurrentPageNumber(val, updateInProgress) {
```
I'm wondering if we actually still need the `updateInProgress` parameter here, and in the events?
Previously we called `scrollPageIntoView` from the event handler in `app.js` (or `viewer.js` as it were), which seemed bad for e.g. the viewer components. In PR #6140, I removed some dead code, and also refactored this. With the changes in this patch, it seems to me at least, that `updateInProgress` is now pointless.
(Obviously there might be someone relying on it, but given that we've not really made any promises about the stability of the viewer components, I think we can remove it.)

yurydelendik replied:
Agree.

yurydelendik committed:
Disable pagechange event if page is not changing.

timvandermeij commented:
/botio-linux preview

Snuffleupagus commented:
I can confirm that issue I reported in https://github.com/mozilla/pdf.js/pull/7289#issuecomment-216791562 is now fixed.

Looks good, thank you for the patch!
